duckdb-vortex/src/include/vortex_common.hpp

```diff
@@ -40,106 +37,100 @@ struct DType {
+struct ConversionCache {
+	explicit ConversionCache(const unsigned long cache_id) : cache(vx_conversion_cache_create(cache_id)) {
 	}
 
-	~VortexConversionCache() {
+	~ConversionCache() {
 		vx_conversion_cache_free(cache);
 	}
 
 	vx_conversion_cache *cache;
 };
 
-struct VortexFileReader {
-	explicit VortexFileReader(vx_file_reader *file) : file(file) {
+struct FileReader {
+	explicit FileReader(vx_file_reader *file) : file(file) {
 	}
 
-	~VortexFileReader() {
+	~FileReader() {
 		vx_file_reader_free(file);
 	}
 
-	static duckdb::unique_ptr<VortexFileReader> Open(const vx_file_open_options *options) {
-		vx_error *error;
-		auto file = vx_file_open_reader(options, &error);
-		HandleError(error);
-		return duckdb::make_uniq<VortexFileReader>(file);
+	static duckdb::unique_ptr<FileReader> Open(const vx_file_open_options *options) {
+		auto file = Try([&](auto err) { return vx_file_open_reader(options, err); });
+		return duckdb::make_uniq<FileReader>(file);
 	}
 
 	uint64_t FileRowCount() {
 		return Try([&](auto err) { return vx_file_row_count(file, err); });
 	}
 
 	struct DType DType() {
-		return ::DType(vx_file_dtype(file));
+		return vortex::DType(vx_file_dtype(file));
 	}
 
 	vx_file_reader *file;
 };
 
-struct VortexArray {
-	explicit VortexArray(vx_array *array) : array(array) {
+struct Array {
+	explicit Array(vx_array *array) : array(array) {
 	}
 
-	~VortexArray() {
+	~Array() {
 		if (array != nullptr) {
 			vx_array_free(array);
 		}
 	}
 
-	static duckdb::unique_ptr<VortexArray> FromDuckDBChunk(DType &dtype, duckdb::DataChunk &chunk) {
-		vx_error *error;
-		auto array = vx_duckdb_chunk_to_array(reinterpret_cast<duckdb_data_chunk>(&chunk), dtype.dtype, &error);
-		HandleError(error);
-		return duckdb::make_uniq<VortexArray>(array);
+	static duckdb::unique_ptr<Array> FromDuckDBChunk(DType &dtype, duckdb::DataChunk &chunk) {
+		auto array = Try([&](auto err) {
+			return vx_duckdb_chunk_to_array(reinterpret_cast<duckdb_data_chunk>(&chunk), dtype.dtype, err);
+		});
+
+		return duckdb::make_uniq<Array>(array);
 	}
 
-	idx_t ToDuckDBVector(idx_t current_row, duckdb_data_chunk output, const VortexConversionCache *cache) const {
-		vx_error *error;
-		auto idx = vx_array_to_duckdb_chunk(array, current_row, output, cache->cache, &error);
-		HandleError(error);
-		return idx;
+	idx_t ToDuckDBVector(idx_t current_row, duckdb_data_chunk output, const ConversionCache *cache) const {
+		return Try([&](auto err) { return vx_array_to_duckdb_chunk(array, current_row, output, cache->cache, err); });
 	}
 
 	vx_array *array;
 };
 
-struct VortexArrayStream {
-	explicit VortexArrayStream(vx_array_stream *array_stream) : array_stream(array_stream) {
+struct ArrayIterator {
+	explicit ArrayIterator(vx_array_iter *array_iter) : array_iter(array_iter) {
 	}
 
-	~VortexArrayStream() {
-		vx_array_stream_free(array_stream);
+	~ArrayIterator() {
+		vx_array_iter_free(array_iter);
```

Comment:
What are the semantics for freeing an iterator if you've consumed all the entries from it? What would be considered standard?

Reply:
Yes, we free the iterator if all elements from that iterator are consumed. The heap allocation is really only needed to share the iterator across FFI.